File: lanscape-agent/internal/agent/websocket.go
package agent

import (
	"context"
	"log/slog"
	"net/http"
	"sync"
	"time"

	"github.com/jhead/lanscape/lanscape-agent/pkg/protocol"
	"nhooyr.io/websocket"
	"nhooyr.io/websocket/wsjson"
)

// WebSocketServer handles browser WebSocket connections
type WebSocketServer struct {
	addr            string
	signalingURL    string
	topic           string
	tailscaleInfo   *TailscaleInfo
	logger          *slog.Logger
	server          *http.Server
	sessions        map[*websocket.Conn]*BrowserSession
	mu              sync.RWMutex
}

// NewWebSocketServer creates a new WebSocket server
func NewWebSocketServer(addr, signalingURL, topic string, tailscaleInfo *TailscaleInfo, logger *slog.Logger) *WebSocketServer {
	return &WebSocketServer{
		addr:          addr,
		signalingURL:  signalingURL,
		topic:         topic,
		tailscaleInfo: tailscaleInfo,
		logger:        logger,
		sessions:      make(map[*websocket.Conn]*BrowserSession),
	}
}

// Start starts the WebSocket server
func (s *WebSocketServer) Start() error {
	mux := http.NewServeMux()
	mux.HandleFunc("/", s.handleWebSocket)

	s.server = &http.Server{
		Addr:    s.addr,
		Handler: mux,
	}

	s.logger.Info("starting WebSocket server", "addr", s.addr)
	return s.server.ListenAndServe()
}

// Stop stops the WebSocket server
func (s *WebSocketServer) Stop(ctx context.Context) error {
	s.mu.Lock()
	for conn, session := range s.sessions {
		session.Disconnect()
		conn.Close(websocket.StatusNormalClosure, "server shutting down")
	}
	s.mu.Unlock()

	if s.server != nil {
		return s.server.Shutdown(ctx)
	}
	return nil
}

// handleWebSocket handles a WebSocket connection
func (s *WebSocketServer) handleWebSocket(w http.ResponseWriter, r *http.Request) {
	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{
		OriginPatterns: []string{"*"}, // Allow all origins for localhost
	})
	if err != nil {
		s.logger.Error("failed to accept WebSocket", "error", err)
		return
	}

	// Create a new browser session for this connection
	session, err := NewBrowserSession(s.signalingURL, s.topic, s.tailscaleInfo, s.logger)
	if err != nil {
		s.logger.Error("failed to create browser session", "error", err)
		conn.Close(websocket.StatusInternalError, "failed to create session")
		return
	}

	// Set up bridge to send messages to this browser (before connecting)
	bridge := session.GetBridge()
	bridge.SetBrowserSend(func(msg protocol.AgentMessage) error {
		return s.sendToBrowser(conn, msg)
	})

	// Connect to signaling server
	if err := session.Connect(); err != nil {
		s.logger.Error("failed to connect to signaling", "error", err)
		conn.Close(websocket.StatusInternalError, "failed to connect to signaling")
		return
	}

	s.mu.Lock()
	s.sessions[conn] = session
	s.mu.Unlock()

	// Wait a bit for welcome message from signaling
	// The signaling client will receive welcome and set selfID
	// We'll send welcome to browser when we receive it from signaling
	// For now, just log
	s.logger.Info("browser connected, waiting for signaling welcome")

	// Handle messages from browser
	ctx := r.Context()
	for {
		var msg protocol.BrowserMessage
		if err := wsjson.Read(ctx, conn, &msg); err != nil {
			s.logger.Debug("browser disconnected", "error", err)
			break
		}

		s.logger.Info("received browser message", "type", msg.Type, "peerId", msg.PeerID, "dataSize", len(msg.Data))

		if err := bridge.HandleBrowserMessage(msg); err != nil {
			s.logger.Warn("failed to handle browser message", "error", err)
			s.sendError(conn, err.Error())
		}
	}

	s.mu.Lock()
	session.Disconnect()
	delete(s.sessions, conn)
	s.mu.Unlock()

	conn.Close(websocket.StatusNormalClosure, "")
	s.logger.Info("browser disconnected")
}

// sendToBrowser sends a message to the browser
func (s *WebSocketServer) sendToBrowser(conn *websocket.Conn, msg protocol.AgentMessage) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	// Always send as JSON (data will be encoded as array)
	return wsjson.Write(ctx, conn, msg)
}

// sendError sends an error message to the browser
func (s *WebSocketServer) sendError(conn *websocket.Conn, errorMsg string) {
	msg := protocol.AgentMessage{
		Type:  protocol.MessageTypeError,
		Error: errorMsg,
	}
	s.sendToBrowser(conn, msg)
}

File: packages/tauri/src-tauri/agent/src/websocket.rs
use crate::protocol::{AgentMessage, BrowserMessage};
use crate::session::BrowserSession;
use crate::tailscale::TailscaleInfo;
use anyhow::{Context, Result};
use axum::{
    extract::{
        ws::{Message, WebSocket, WebSocketUpgrade},
        State,
    },
    response::Response,
    routing::get,
    Router,
};
use std::collections::HashMap;
use std::net::SocketAddr;
use std::sync::Arc;
use tokio::sync::RwLock;
use tracing::{debug, error, info, warn};

/// WebSocketServer handles browser WebSocket connections
pub struct WebSocketServer {
    pub addr: String,
    signaling_url: String,
    topic: String,
    tailscale_info: Option<TailscaleInfo>,
    sessions: Arc<RwLock<HashMap<String, Arc<BrowserSession>>>>,
}

impl WebSocketServer {
    /// Create a new WebSocket server
    pub fn new(
        addr: String,
        signaling_url: String,
        topic: String,
        tailscale_info: Option<TailscaleInfo>,
    ) -> Self {
        Self {
            addr,
            signaling_url,
            topic,
            tailscale_info,
            sessions: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    /// Start the WebSocket server
    pub async fn start(&self) -> Result<()> {
        let sessions = self.sessions.clone();
        let signaling_url = self.signaling_url.clone();
        let topic = self.topic.clone();
        let tailscale_info = self.tailscale_info.clone();

        let app = Router::new()
            .route("/", get(handle_websocket))
            .with_state(ServerState {
                sessions: sessions.clone(),
                signaling_url: signaling_url.clone(),
                topic: topic.clone(),
                tailscale_info: tailscale_info.clone(),
            });

        let addr: SocketAddr = self.addr.parse().context("Invalid address")?;
        info!("Starting WebSocket server: addr={}", addr);

        let listener = tokio::net::TcpListener::bind(addr)
            .await
            .context("Failed to bind to address")?;

        axum::serve(listener, app)
            .await
            .context("Server error")?;

        Ok(())
    }

    /// Stop the WebSocket server
    pub async fn stop(&self) -> Result<()> {
        let mut sessions = self.sessions.write().await;
        for (_, session) in sessions.drain() {
            if let Err(e) = session.stop().await {
                warn!("Error stopping session: {}", e);
            }
        }
        Ok(())
    }
}

#[derive(Clone)]
struct ServerState {
    sessions: Arc<RwLock<HashMap<String, Arc<BrowserSession>>>>,
    signaling_url: String,
    topic: String,
    tailscale_info: Option<TailscaleInfo>,
}

async fn handle_websocket(
    ws: WebSocketUpgrade,
    State(state): State<ServerState>,
) -> Response {
    ws.on_upgrade(move |socket| handle_socket(socket, state))
}

async fn handle_socket(socket: WebSocket, state: ServerState) {
    let (sender, mut receiver) = socket.split();
    let session_id = uuid::Uuid::new_v4().to_string();

    // Create a new browser session for this connection
    let session = match BrowserSession::new(
        state.signaling_url.clone(),
        state.topic.clone(),
        state.tailscale_info.clone(),
    )
    .await
    {
        Ok(s) => Arc::new(s),
        Err(e) => {
            error!("Failed to create browser session: {}", e);
            let sender_wrapped = Arc::new(tokio::sync::Mutex::new(sender));
            let mut sender_guard = sender_wrapped.lock().await;
            let _ = sender_guard
                .send(Message::Close(Some(axum::extract::ws::CloseFrame {
                    code: 1011u16.into(),
                    reason: "Failed to create session".into(),
                })))
                .await;
            return;
        }
    };

    // Set up bridge to send messages to this browser
    let bridge = session.get_bridge();
    let sender = Arc::new(tokio::sync::Mutex::new(sender));
    let sender_for_bridge = sender.clone();
    bridge
        .set_browser_send(Box::new(move |msg: AgentMessage| -> Result<()> {
            let sender = sender_for_bridge.clone();
            tokio::spawn(async move {
                match serde_json::to_string(&msg) {
                    Ok(text) => {
                        let mut sender_guard = sender.lock().await;
                        if let Err(e) = sender_guard.send(Message::Text(text)).await {
                            error!("Failed to send message to browser: {}", e);
                        }
                    }
                    Err(e) => {
                        error!("Failed to serialize message: {}", e);
                    }
                }
            });
            Ok(())
        }))
        .await;

    // Connect to signaling server
    if let Err(e) = session.connect().await {
        error!("Failed to connect to signaling: {}", e);
        let mut sender_guard = sender.lock().await;
        let _ = sender_guard
            .send(Message::Close(Some(axum::extract::ws::CloseFrame {
                code: 1011u16.into(),
                reason: "Failed to connect to signaling".into(),
            })))
            .await;
        return;
    }

    state
        .sessions
        .write()
        .await
        .insert(session_id.clone(), session.clone());

    let sender_for_errors = sender.clone();
    info!("Browser connected, waiting for signaling welcome");

    // Handle messages from browser
    while let Some(msg) = receiver.next().await {
        match msg {
            Ok(Message::Text(text)) => {
                match serde_json::from_str::<BrowserMessage>(&text) {
                    Ok(browser_msg) => {
                        if let Err(e) = bridge.handle_browser_message(browser_msg).await {
                            warn!("Failed to handle browser message: {}", e);
                            let error_msg = AgentMessage::error(e.to_string());
                            if let Ok(text) = serde_json::to_string(&error_msg) {
                                let mut sender_guard = sender_for_errors.lock().await;
                                let _ = sender_guard.send(Message::Text(text)).await;
                            }
                        }
                    }
                    Err(e) => {
                        warn!("Failed to parse browser message: {}", e);
                    }
                }
            }
            Ok(Message::Close(_)) => {
                debug!("Browser disconnected");
                break;
            }
            Err(e) => {
                debug!("WebSocket error: {}", e);
                break;
            }
            _ => {}
        }
    }

    state.sessions.write().await.remove(&session_id);
    session.disconnect().await;
    info!("Browser disconnected");
}

use futures_util::{SinkExt, StreamExt};

